Add support for scheduled events

jamesls opened the pull request:
Proposal: https://github.com/awslabs/chalice/issues/390

Sample app:

```python
from chalice import Chalice
import datetime

app = Chalice(app_name='testevents')


@app.route('/')
def index():
    return {'hello': 'world'}


@app.schedule('rate(1 minute)')
def every_minute(event):
    print(event.to_dict())

```


The implementation piggy backs off of the multi lambda support from the built in authorizers.  The only unexpected change was an update to the deployed.json file from:

```
    "lambda_functions": {
      "testevents-dev-every_minute": "arn:lambda-foo"
    },

```
to:

```
    "lambda_functions": {
      "testevents-dev-every_minute": {
        "type": "scheduled_event",
        "arn": "arn:lambda-foo"
      }
    },

```

So we can differentiate the different types of lambda functions.

The other caveat is that right now you still need to deploy an API gateway API.  There isn't support yet for only deploying scheduled events.  That will be in a follow up PR.

**Review Notes**

The deployer module has some duplication that I'm working on removing.  The logic itself won't change but I'm refactoring the code.  Everything else (`app.py`, `awsclient.py`, all the tests) are ready for review.

jamesls commented on chalice/app.py:
```diff
@@ -765,3 +779,82 @@ class AuthRoute(object):
+
+class CloudWatchEvent(object):
+    def __init__(self, event_dict):
+        self.version = event_dict['version']
```
Yes.

jamesls commented on chalice/app.pyi:
```diff
@@ -148,3 +149,31 @@ class AuthResponse(object):
+
+
+class Rate(ScheduleExpression):
+    unit = ... # type: str
```
That should be an integer, I'll update.

jamesls commented on chalice/app.py:
```diff
@@ -765,3 +779,82 @@ class AuthRoute(object):
         self.methods = methods
+
+
+class EventSource(object):
```
It's intentional to have an `EventSource` type.  The idea would be this is used for all event types as support more types is added.  I don't envision adding a new public attr to `chalice.Chalice` for every new event source, that wouldn't be maintainable.  I can extract out a separate `CloudWatchEventSource` class with `EventSource` as its base type.  Seemed like overkill for now, but I don't mind adding it.

kyleknap replied:
Makes sense. I like the idea of having `EventSource` be the base class and `CloudWatchEventSource` subclass from it to include any of the cloudwatch event specific logic.

jamesls replied:
I pushed an update.  Let me know what you think.

kyleknap commented on tests/functional/test_awsclient.py:
```diff
@@ -906,17 +906,48 @@ def test_can_add_permission_for_apigateway_not_needed(self, stubbed_session):
+                         'AWS:SourceArn': source_arn,
+                     }
+                 },
+                 'Effect': 'Allow',
```
Without looking at what got removed in the diff, it is not very easy to tell what portion of the policy needs to be present for it to be skipped. Maybe separate this into tests where one tests that a specific policy entry skips over the `add_permission` call but the other entries causes `add_permission` to be triggered?

jamesls replied:
I refactored the test a bit to make this more clear.

kyleknap commented on chalice/app.py:
```diff
@@ -765,3 +779,87 @@ class AuthRoute(object):
+        return 'rate(%s %s)' % (self.value, unit)
+
+
+class Cron(ScheduleExpression):
```
Did we want to add `Cron` and `Rate` to the `__init__`?

jamesls replied:
I don't follow what you're asking.  Can you elaborate?

kyleknap replied:
Oh so I have to import those two from `chalice.app` right now, but I'm use to importing all of the user facing classes from `chalice` instead when I right my application.

jamesls replied:
ohhhh, `chalice/__init__.py`.  Gotcha.  Yeah I can add that.

kyleknap commented:
Cool. Looks good once docs are included, it should be good to merge. I also noticed that there was some missing coverage according to the codecov report. Do you know what that may be?

jamesls commented:
Yeah it was some of the new code I added for the base class refactoring.  I added a `isinstance` check that's missing a testcase if you don't pass an instance of `CloudWatchEventSource`.  Right now there are no other kinds so I can make one up for the test.

jamesls commented:
OK I think everything has been addressed, and docs have been added.  Let me know if there's any more feedback.

cc @kyleknap @stealthycoin

jamesls commented on docs/source/api.rst:
```diff
@@ -118,6 +118,46 @@ Chalice
+      .. code-block:: python
+
+          @app.schedule('cron(15 10 ? * 6L 2002-2005)')
+          def handler(event):
```
You mean not share?

kyleknap replied:
yes not share

jamesls replied:
OK I'll update that before merging.
